rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RetryingBookkeeperClient.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public final class RetryingBookkeeperClient
+        extends BookKeeperService.Client
+        implements Closeable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(RetryingBookkeeperClient.class);
+    private int maxRetries;
+    TTransport transport;
+
+    public RetryingBookkeeperClient(TTransport transport, int maxRetries)
+    {
+        super(new TBinaryProtocol(transport));
+        this.transport = transport;
+        this.maxRetries = maxRetries;
+    }
+
+    public static RetryingBookkeeperClient createBookKeeperClient(Configuration conf)
+            throws TTransportException
+    {
+        TTransport transport;
+        transport = new TSocket("localhost", CacheConfig.getServerPort(conf), CacheConfig.getClientTimeout(conf));
+        transport.open();
+
+        RetryingBookkeeperClient retryingBookkeeperClient = new RetryingBookkeeperClient(transport, CacheConfig.getMaxRetries(conf));
+        return retryingBookkeeperClient;
+    }
+
+    @Override
+    public List<Location> getCacheStatus(final String remotePath, final long fileLength, final long lastModified, final long startBlock, final long endBlock, final int clusterType)
+            throws TException
+    {
+        return retryConnection(new Callable<List<Location>>()
+        {
+            @Override
+            public List<Location> call()
+                    throws TException
+            {
+                return RetryingBookkeeperClient.super.getCacheStatus(remotePath, fileLength, lastModified, startBlock, endBlock, clusterType);
+            }
+        });
+    }
+
+    @Override
+    public void setAllCached(final String remotePath, final long fileLength, final long lastModified, final long startBlock, final long endBlock)
+            throws TException
+    {
+        retryConnection(new Callable<Void>()
+        {
+            @Override
+            public Void call()
+                    throws Exception
+            {
+                RetryingBookkeeperClient.super.setAllCached(remotePath, fileLength, lastModified, startBlock, endBlock);
+                return null;
+            }
+        });
+    }
+
+    //Assuming that transport from bookKeeperClient is already open
+    private <V> V retryConnection(Callable<V> callable)
+            throws TException
+    {
+        int errors = 0;
+
+        while (errors < maxRetries) {
+            try {
+                if (!transport.isOpen()) {
+                    try {
+                        transport.open();
+                    }
+                    catch (Exception e1) {
+                        LOG.info("Error while reconnecting");
```

Comment:
this could end up in infinite loop, let the transport.open's exception get caught outside so that errors are incremented and we can break